File: ocaml/lib/day05.ml
type page = int

type rule = 
    { before: page
    ; after: page
    }

module Update = struct
    type t = page list

    let of_list xs : t = xs 

    let is_ordered t rules = 
        let eval_rule rule = 
            let open CCOption.Infix in
            let* p1 = List.find_index ((=) rule.before) t in
            let* p2 = List.find_index ((=) rule.after) t in
            Some (p1 < p2)
        in
        rules |> List.filter_map(eval_rule) |> CCList.for_all (fun x -> x)

    let middle t = CCList.get_at_idx_exn ((List.length t) / 2) t
end

type problem = 
    { rules: rule list
    ; updates: Update.t list
    }

module Parsers = struct
    open CCParse
    open CCParse.Infix

    let rule =
        return (fun before after -> { before; after }) <*> U.int <*> char '|' *> U.int

    let update = sep ~by:(char ',') U.int |> map Update.of_list
    
    let problem =
        let* rules = sep ~by:(endline) rule in
        let* _ = endline in
        let* updates = 
            sep ~by:(endline) update
            |> map (CCList.filter (fun u -> List.length u > 0))
        in
        return { rules ; updates }
end

let part1 () = 
    let problem = Aoclib.parse_input_exn Parsers.problem in
    problem.updates
    |> CCList.filter_map (fun u -> 
        if Update.is_ordered u problem.rules then Some(Update.middle u) else None
    )
    |> List.fold_left ( + ) 0
    |> Printf.printf "%d\n"

File: rust/aoc24/src/bin/day05.rs
type Page = u32;

#[derive(Debug)]
struct Rule {
    before: Page,
    after: Page,
}

#[derive(Debug, Clone)]
struct Update {
    pages: Vec<Page>,
}

#[derive(Debug)]
struct Problem {
    rules: Vec<Rule>,
    updates: Vec<Update>,
}

impl Update {
    pub fn is_ordered(&self, rules: &Vec<Rule>) -> bool {
        rules
            .iter()
            .filter_map(|rule| {
                let p1 = self.pages.iter().position(|p| *p == rule.before)?;
                let p2 = self.pages.iter().position(|p| *p == rule.after)?;
                Some(p1 < p2)
            })
            .all(|v| v == true)
    }

    pub fn middle(&self) -> Page {
        self.pages[self.pages.len() / 2]
    }
}

mod parsers {
    use super::{Problem, Rule, Update};
    use nom::{
        bytes::complete::tag, character::complete::*, multi::*, sequence::*, IResult, Parser,
    };

    fn rule(input: &str) -> IResult<&str, Rule> {
        separated_pair(u32, tag("|"), u32)
            .map(|v| Rule {
                before: v.0,
                after: v.1,
            })
            .parse(input)
    }

    fn update(input: &str) -> IResult<&str, Update> {
        separated_list0(tag(","), u32)
            .map(|v| Update { pages: v })
            .parse(input)
    }

    pub fn problem<'a>(input: &'a str) -> IResult<&'a str, Problem> {
        let rules = |s: &'a str| separated_list0(newline, rule).parse(s);
        let updates = |s: &'a str| separated_list0(newline, update).parse(s);
        separated_pair(rules, pair(newline, newline), updates)
            .map(|v| Problem {
                rules: v.0,
                updates: v
                    .1
                    .iter()
                    .filter(|v| v.pages.is_empty() == false)
                    .cloned()
                    .collect(),
            })
            .parse(input)
    }
}

fn main() {
    let input = aoc24::aoclib::parse_all_stdin(parsers::problem);
    let answer: u32 = input
        .updates
        .iter()
        .filter_map(|u| {
            if u.is_ordered(&input.rules) {
                Some(u.middle())
            } else {
                None
            }
        })
        .sum();
    println!("{answer}");
}
